src/recon.rs

```diff
@@ -817,33 +840,75 @@ fn decode_coefs<BD: BitDepth>(
         let shift;
         let shift2;
         let mask;
-        let slwh_zero;
+        let end: usize;
         match tx_class {
             TxClass::TwoD => {
                 shift = slh + 2;
                 shift2 = 0;
                 mask = (4 << slh) - 1;
-                slwh_zero = slw;
+                // We need to zero part of `levels`. We want to calulate the end
+                // of the slice to zero in a way that rust can elide the bounds
+                // check. The size of the `levels` array is 1088 bytes.
+                //
+                // The original calculation here:
+                // `end = stride as usize * ((4 << slw) as usize + 2)`
+                //
+                // We know:
+                //  - `stride = 4 << slh`, where `slh <= 3`
+                //    (therefore `stride <= 32`)
+                //
+                //  - `slw <= 3` therefore `4 << slw <= 32`
+                //
+                // Therfore, we have the maximum `end` value as
+                //  `32 * (32 + 2) = 1088`
+                // Therefore, no bounds check should be required to slice `levels`.
+                //
+                // However, as of nightly 2025-05-01, rustc had a lot of trouble
+                // proving that to itself. Here we give it a hand. First, we
```

Comment:
Even with extra `assert!`s, it can't do this?  That's surprising.  Usually LLVM is better than that.

Reply:
Oh sorry, I totally misunderstood the context of your comment. With the rearranged calculation , the asserts on `term1, `term2, and the final bounds check before `fill` are all elided. The only thing that I can't get regardless of asserts is the addition vs bitwise or.